Mathlib/RingTheory/Coalgebra.lean

```diff
@@ -0,0 +1,64 @@
+/-
+Copyright (c) 2023 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Definition and example of a coalgebra
+-/
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A coalgebra over a commutative ring `R` is a module over `R` equipped with a coassociative
+comultiplication `Δ` and a counit `ε` obeying the left and right conunitality laws. -/
+class Coalgebra (R : Type u) (A : Type v) [CommRing R] [AddCommGroup A] [Module R A] where
+  /-- The comultiplication of the coalgebra -/
+  Δ : A →ₗ[R] A ⊗[R] A
+  /-- The counit of the coalgebra -/
+  ε : A →ₗ[R] R
+  /-- The comultiplication is coassociative -/
+  coassoc : TensorProduct.assoc R A A A ∘ₗ TensorProduct.map Δ .id ∘ₗ Δ =
+    TensorProduct.map .id Δ ∘ₗ Δ
+  /-- The counit satisfies the left counitality law -/
+  ε_id : TensorProduct.lid R A ∘ₗ TensorProduct.map ε .id ∘ₗ Δ = .id
+  /-- The counit satisfies the right counitality law -/
+  id_ε : TensorProduct.rid R A ∘ₗ TensorProduct.map .id ε ∘ₗ Δ = .id
+
+namespace Coalgebra
+
+@[simp]
+theorem coassoc_apply {R : Type u} {A : Type v} [CommRing R] [AddCommGroup A] [Module R A]
+    [C : Coalgebra R A] (a : A) : TensorProduct.assoc R A A A (TensorProduct.map C.Δ .id (C.Δ a)) =
```

Comment:
This raises that something a big ugly is going on here; for algebras, we wouldn't usually write `Alg.algebraMap r` (where `Alg : Algebra R A`), we write `algebraMap R A` instead.

One (ugly) option would be to repeat this pattern, and rename `Δ` to `Δ'` so that you can then `def Δ := Δ'`, but with an explicit `R` variable.

You might also be able to get away with `Δ (R := R)` instead of `C.Δ` without changing anything else, which we didn't have as an option when `Algebra` was designed.

Reply:
I think since I added the `namespace Coalgebra` I can replace `C : Coalgebra R A` with just `Coalgebra R A` and `C.Δ` with just `Δ` and it seems to work fine? But then in general, if I had theorems involving `[Coalgebra R A]` outside of the namespace, would I want to have 
```
def Δ (R : Type u) {A : Type v} [CommRing R] [Ring A] [Module R A] [C : Coalgebra R A] := C.Δ
```
just before them? (I did try `Δ (R := R)` but I couldn't get it working.)

Reply:
I don'f fully understand what you mean, but if you have a version with fewer `C.`s, feel free to push it.

Reply:
Ah, now I see: this is great, but I think the comment above will bite you eventually. In your case, Lean can work out `R` because it's an argument to `TensorProduct.assoc`. In general though, Lean might force you to write extra type annotations.

I think it's reasonable to leave it as is for now; but I predict you will end up running into this problem again in a future PR, and at that point we should revisit it.